File: java/jimmer-simple/src/main/java/org/babyfish/jimmer/sql/example/web/DtoStyleController.java
package org.babyfish.jimmer.sql.example.web;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.example.model.*;
import org.babyfish.jimmer.sql.example.model.dto.ComplexBookView;
import org.babyfish.jimmer.sql.example.model.dto.RecursiveTreeNodeView;
import org.babyfish.jimmer.sql.example.model.dto.SimpleBookView;
import org.babyfish.jimmer.sql.example.model.dto.SimpleTreeNodeView;
import org.jetbrains.annotations.Nullable;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/dto")
@Transactional
public class DtoStyleController implements Tables {

    private final JSqlClient sqlClient;

    public DtoStyleController(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    /**
     * Find simple book objects.
     *
     * <p>If this parameter is specified, only books with the specified name will be queried.</p>
     *
     * @param name Optional query parameter,
     *      {@code null} or {@code ""} will be ignored
     * @return A list of Book objects, each of which
     *      has only {@code id} and {@code name} properties
     */
    @GetMapping("/books/simple")
    public List<SimpleBookView> findSimpleBooks(
            @Nullable @RequestParam(required = false) String name
    ) {
        BookTable table = BOOK_TABLE;
        return sqlClient.createQuery(table)
                .where(table.name().eqIf(name))
                .select(table.fetch(SimpleBookView.class))
                .execute();
    }

    /**
     * Find complex book objects.
     *
     * <p>If this parameter is specified, only books with the specified name will be queried.</p>
     *
     * @param name Optional query parameter,
     *      {@code null} or {@code ""} will be ignored
     * @return A list of Book objects, each of which
     *      has all scalar properties, associated
     *      {@link BookStore} and associated
     *      {@link Author} objects.
     */
    @GetMapping("/books/complex")
    public List<ComplexBookView> findComplexBooks(
            @Nullable @RequestParam(required = false) String name
    ) {
        BookTable table = BOOK_TABLE;
        return sqlClient.createQuery(table)
                .where(table.name().eqIf(name))
                .select(table.fetch(ComplexBookView.class))
                .execute();
    }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     *         has only {@code id} and {@code name} properties
     */
    @GetMapping("/rootTreeNodes/simple")
    public List<SimpleTreeNodeView> findSimpleTreeNodes() {
        TreeNodeTable table = TREE_NODE_TABLE;
        return sqlClient.createQuery(table)
                .where(table.parentId().isNull())
                .select(table.fetch(SimpleTreeNodeView.class))
                .execute();
    }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     *         has all scalar properties, associated child objects.
     *         If a child node has deeper child nodes,
     *         it will be recursively associated until
     *         there are no deeper child nodes.
     */
    @GetMapping("/rootTreeNodes/recursive")
    public List<RecursiveTreeNodeView> findRecursiveTreeNodes() {
        TreeNodeTable table = TREE_NODE_TABLE;
        return sqlClient.createQuery(table)
                .where(table.parentId().isNull())
                .select(table.fetch(RecursiveTreeNodeView.class))
                .execute();
    }
}

File: kotlin/jimmer-simple-kt/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/web/DtoStyleController.kt
package org.babyfish.jimmer.sql.example.kt.web

import org.babyfish.jimmer.sql.example.kt.model.*
import org.babyfish.jimmer.sql.example.kt.model.dto.ComplexBookView
import org.babyfish.jimmer.sql.example.kt.model.dto.RecursiveTreeNodeView
import org.babyfish.jimmer.sql.example.kt.model.dto.SimpleBookView
import org.babyfish.jimmer.sql.example.kt.model.dto.SimpleTreeNodeView
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.`eq?`
import org.babyfish.jimmer.sql.kt.ast.expression.isNull
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/dto")
@Transactional
class DtoStyleController(
    private val sqlClient: KSqlClient
) {

    /**
     * Find simple book objects.
     *
     * If this parameter is specified, only books with the specified name will be queried.
     *
     * @param name Optional query parameter,
     * `null` and `""` will be ignored
     *
     * @return A list of Book objects, each of which
     * has only `name` and `edition` properties
     */
    @GetMapping("/books/simple")
    fun findSimpleBooks(
        @RequestParam(required = false) name: String?
    ): List<SimpleBookView> =
        sqlClient.executeQuery(Book::class) {
            where(table.name `eq?` name)
            select(table.fetch(SimpleBookView::class))
        }

    /**
     * Find complex book objects.
     *
     * If this parameter is specified, only books with the specified name will be queried.
     *
     * @param name Optional query parameter,
     * `null` and `""` will be ignored
     *
     * @return A list of Book objects, each of which
     * has all scalar properties, associated
     * [BookStore] and associated
     * [Author] objects.
     */
    @GetMapping("/books/complex")
    fun findComplexBooks(
        @RequestParam(required = false) name: String?
    ): List<ComplexBookView> =
        sqlClient.executeQuery(Book::class) {
            where(table.name `eq?` name)
            select(table.fetch(ComplexBookView::class))
        }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     * has only `id` and `name` properties
     */
    @GetMapping("/rootTreeNodes/simple")
    fun findSimpleTreeNodes(
    ): List<SimpleTreeNodeView> =
        sqlClient.executeQuery(TreeNode::class) {
            where(table.parentId.isNull())
            select(table.fetch(SimpleTreeNodeView::class))
        }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     *         has all scalar properties, associated child objects.
     *         If a child node has deeper child nodes,
     *         it will be recursively associated until
     *         there are no deeper child nodes.
     */
    @GetMapping("/rootTreeNodes/recursive")
    fun findRecursiveTreeNodes(
    ): List<RecursiveTreeNodeView> =
        sqlClient.executeQuery(TreeNode::class) {
            where(table.parentId.isNull())
            select(table.fetch(RecursiveTreeNodeView::class))
        }
}